src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/InternalEvaluatorImpl.kt

```diff
@@ -124,4 +158,73 @@ internal class InternalEvaluatorImpl(
             isExecuting = false
         }
     }
+
+    private fun updateVariablesState(cellId: Int) {
+        // remove known modifying usages in this cell
+        usageMap[cellId]?.removeIf {
+            containingMap[it] != cellId
+        }
+
+        variablesMap.forEach {
+            val state = it.value
+            val property = state.propertyKlass
+            val newValue = property.get(state.scriptInstance) ?: return@forEach
+
+            if (state.stringValue != newValue.toString()) {
+                usageMap[cellId]?.add(it.key)
+                state.stringValue = newValue.toString()
+            }
+        }
+    }
+
+    private fun getVisibleVariables(target: ResultValue, cellId: Int): Map<String, VariableState> {
+        val klass = target.scriptClass ?: return emptyMap()
+        val cellKlassInstance = target.scriptInstance!!
+
+        val fields = klass.declaredMemberProperties
+        val ans = mutableMapOf<String, VariableState>()
+        fields.forEach {
+            val casted = it as KProperty1<Any, *>
+            ans[casted.name] = if (casted.get(cellKlassInstance) == null) {
```

Comment:
Why variable value can't be null?

Reply:
Fair enough)

Reply:
I'm thinking about this invariant: all process is here, so we can literally send "null" as a string value. Would it be an overhead? Instead, we could send just String -> String?, and not overload the channel with needless bytes

Reply:
I think that "null" is a legit value for user to see. I don't think that we should optimize something here